sensing/autoware_crop_box_filter/include/autoware/crop_box_filter/crop_box_filter_node.hpp

```diff
@@ -0,0 +1,189 @@
+    float min_y;
+    float max_y;
+    float min_z;
+    float max_z;
+    bool negative{false};
+  } param_;
+
+  /** \brief Parameter service callback result : needed to be hold */
+  OnSetParametersCallbackHandle::SharedPtr set_param_res_;
+
+  // publisher and subscriber declaration *********************
+
+  /** \brief The input PointCloud2 subscriber. */
+  rclcpp::Subscription<PointCloud2>::SharedPtr sub_input_;
+
+  rclcpp::Publisher<PointCloud2>::SharedPtr pub_output_;
+  rclcpp::Publisher<geometry_msgs::msg::PolygonStamped>::SharedPtr crop_box_polygon_pub_;
+
+  /** \brief processing time publisher. **/
+  std::unique_ptr<autoware_utils::StopWatch<std::chrono::milliseconds>> stop_watch_ptr_;
+  std::unique_ptr<autoware_utils::DebugPublisher> debug_publisher_;
+  std::unique_ptr<autoware_utils::PublishedTimePublisher> published_time_publisher_;
+
+  // function declaration *************************************
+
+  void publish_crop_box_polygon();
+
+  void pointcloud_callback(const PointCloud2ConstPtr cloud);
+
+  /** \brief Parameter service callback */
+  rcl_interfaces::msg::SetParametersResult param_callback(const std::vector<rclcpp::Parameter> & p);
+
+  /** \brief Return whether the input PointCloud2 data has the same layout than PointXYZI. That is
+   * to say whether you can memcpy from the PointCloud2 data buffer to a PointXYZI */
+  bool is_data_layout_compatible_with_point_xyzi(const PointCloud2 & input);
+
+  /** \brief Return whether the input PointCloud2 data has the same layout than PointXYZIRC. That is
+   * to say whether you can memcpy from the PointCloud2 data buffer to a PointXYZIRC */
+  bool is_data_layout_compatible_with_point_xyzirc(const PointCloud2 & input);
+
+  /** \brief Return whether the input PointCloud2 data has the same layout than PointXYZIRADRT. That
+   * is to say whether you can memcpy from the PointCloud2 data buffer to a PointXYZIRADRT */
+  bool is_data_layout_compatible_with_point_xyziradrt(const PointCloud2 & input);
+
+  /** \brief Return whether the input PointCloud2 data has the same layout than PointXYZIRCAEDT.
+   * That is to say whether you can memcpy from the PointCloud2 data buffer to a PointXYZIRCAEDT */
+  bool is_data_layout_compatible_with_point_xyzircaedt(const PointCloud2 & input);
+
+  bool is_valid(const PointCloud2ConstPtr & cloud);
+
+  /** \brief For parameter service callback */
+  template <typename T>
+  bool get_param(const std::vector<rclcpp::Parameter> & p, const std::string & name, T & value)
+  {
+    auto it = std::find_if(p.cbegin(), p.cend(), [&name](const rclcpp::Parameter & parameter) {
+      return parameter.get_name() == name;
+    });
+    if (it != p.cend()) {
+      value = it->template get_value<T>();
+      return true;
+    }
+    return false;
+  }
+
+public:
+  PCL_MAKE_ALIGNED_OPERATOR_NEW
+  explicit CropBoxFilter(const rclcpp::NodeOptions & options);
+  void pointcloud_filter(const PointCloud2ConstPtr & cloud, PointCloud2 & output);
```

Comment:
Function names should be verb phrases. I think `filter_pointcloud` or `filter` would be correct.
```suggestion
  void filter(const PointCloud2ConstPtr & cloud, PointCloud2 & output);
```

Reply:
Got it kondo san.
let me double check the code, especially the memory efficiency part.

Reply:
Good afternoon Kondo san @youtalk 

Regarding “memory efficiency”， i picked several code blocks which access memeory

|ID|CODE|COMMENT|
|------|----------|-----------|
|1|``` cpp output.data.resize(cloud->data.size()); ```| this code malloc a memory block which has same size as the input pointcloud, since this node aims at crop self points, it will not waste too much memory(most of the points are outside ego).  Alternatively we can replace “resize” with "push_back() in the loop" but that will bring more cost for frequent asking cpu deal with memory, and every time "data.size" is not sufficient , it will expand the memory size by 150%, this may cause memory waste |
|2| ![图片](https://github.com/user-attachments/assets/fd759387-34b7-4e86-8986-d78f8dbac5c5) | this block picks a point from the input poincloud for filtering, rather than copy a point from the orginal  input poincloud in the loop, we can use function pcl::from_msg  to convert the msg into a pcl::pointcloud and iterate every point in the loop but that bring a accessary memory cost for storing the whole point cloud |
|3|  ![图片](https://github.com/user-attachments/assets/717a3b8b-0ccf-4de9-bcd8-d47bc18f0ff3) | this block add the filter result into output pointcloud consider transformation |

Thanks for your review

Please feel free to leave your further advice

心刚